omplbindings/evaluate.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"math"
+	"strings"
+
+	"github.com/golang/geo/r3"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/spatialmath"
+	"gonum.org/v1/gonum/floats"
+)
+
+const defaultEpsilon = 1e-2
+
+func evaluateSolution(solution [][]float64, sceneNum int) (float64, float64, float64, error) {
+	var l2Score, lineScore, oScore float64
+
+	if err := Init(sceneNum); err != nil {
+		return -1, -1, -1, err
+	}
+	thisFrame := sceneFS.Frame(testArmFrame)
+
+	poseStart, err := thisFrame.Transform(referenceframe.FloatsToInputs(solution[0]))
+	if poseStart == nil || (err != nil && !strings.Contains(err.Error(), referenceframe.OOBErrString)) {
+		return -1, -1, -1, err
+	}
+	poseEnd, err := thisFrame.Transform(referenceframe.FloatsToInputs(solution[len(solution)-1]))
+	if poseEnd == nil || (err != nil && !strings.Contains(err.Error(), referenceframe.OOBErrString)) {
+		return -1, -1, -1, err
+	}
+
+	// For each step
+	for i := 0; i < len(solution)-1; i++ {
+		l2Score += L2Distance(solution[i], solution[i+1])
+
+		// Check linear and orientation excursion every 2 degrees of joint movement
+		nSteps := getSteps(solution[i], solution[i+1])
+		for j := 1; j <= nSteps; j++ {
+			step := referenceframe.InterpolateInputs(
+				referenceframe.FloatsToInputs(solution[i]),
+				referenceframe.FloatsToInputs(solution[i+1]),
+				float64(j)/float64(nSteps),
+			)
+			pose, err := thisFrame.Transform(step)
+			if pose == nil || (err != nil && !strings.Contains(err.Error(), referenceframe.OOBErrString)) {
+				return -1, -1, -1, err
+			}
+			lineScore += distToLine(poseStart.Point(), poseEnd.Point(), pose.Point())
+			oScore += orientScore(poseStart.Orientation(), poseEnd.Orientation(), pose.Orientation())
+		}
+	}
+	totalLineDist := poseStart.Point().Sub(poseEnd.Point()).Norm()
+	return l2Score, lineScore / totalLineDist, oScore, nil
+}
+
+// L2Distance returns the L2 normalized difference between two equal length arrays.
+func L2Distance(q1, q2 []float64) float64 {
+	diff := make([]float64, len(q1))
+	for i := 0; i < len(q1); i++ {
+		diff[i] = q1[i] - q2[i]
+	}
+	// 2 is the L value returning a standard L2 Normalization
+	return floats.Norm(diff, 2)
+}
+
+func distToLine(pt1, pt2, query r3.Vector) float64 {
```

Comment:
Could deduplicate this code by using what's in `geometry_utils.go` from the RSDK-251 PR...

Reply:
Will consider this depending on what gets merged first, otherwise will add a TODO